### PR TITLE
feat: add inline edit layer to htmlrender and creator

### PR DESCRIPTION
**Pull request by williamw2**
# Pull Request

## 📖 Description
Adding a new layer to the HTMLRender component that allows for inline editing of text and adding this functionality to Creator. Expected behavior is whenever text within the preview area is double clicked a textarea is placed in the same position of the text node allowing for user input. The text area will resize based on the input and pressing the Enter key will commit the changes. Pressing Esc, Tab or clicking outside of the text area will cancel the interaction.

### General

- [x] I have included a change request file using `$ yarn change`
- [ ] I have added tests for my changes.
- [x] I have tested my changes.
- [ ] I have updated the project documentation to reflect my changes.
- [x] I have read the [CONTRIBUTING](https://github.com/Microsoft/fast/blob/master/CONTRIBUTING.md) documentation and followed the [standards](https://www.fast.design/docs/community/code-of-conduct/#our-standards) for this project.

### Component-specific

- [ ] I have added a new component
- [ ] I have modified an existing component
- [ ] I have updated the [definition file](https://github.com/Microsoft/fast/blob/master/packages/web-components/fast-components/CONTRIBUTING.md#definition)
- [ ] I have updated the [configuration file](https://github.com/Microsoft/fast/blob/master/packages/web-components/fast-components/CONTRIBUTING.md#configuration)

### CODE REVIEW

File: sites/fast-creator/app/web-components/index.tsx [hunk 1]
```diff
@@ -318,6 +320,7 @@ export class HTMLRenderReact extends React.Component {
             <fast-design-system-provider ref={this.setDesignRef}>
                 <fast-tooling-html-render ref={this.setRenderRef}>
                     <fast-tooling-html-render-layer-navigation role="htmlrenderlayer"></fast-tooling-html-render-layer-navigation>
+                    <fast-tooling-html-render-layer-inline-edit role="htmlrenderlayer"></fast-tooling-html-render-layer-inline-edit>
```
Comment (EisenbergEffect) [inline sites/fast-creator/app/web-components/index.tsx:323]: Nifty!

File: packages/tooling/fast-tooling/src/web-components/html-render-layer-inline-edit/html-render-layer-inline-edit.ts [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+        if (this.activityCallback) {
+            this.activityCallback(this.layerActivityId, ActivityType.update);
+        }
+    }
+
+    public handleBlur(e: InputEvent) {
+        this.cancelEdit();
+    }
+
+    private GetPositionFromElement(target: Node): OverylayPosition {
```
Comment (EisenbergEffect) [inline packages/tooling/fast-tooling/src/web-components/html-render-layer-inline-edit/html-render-layer-inline-edit.ts:90]: Just noticed this was capitalized for some reason.

**Review by EisenbergEffect — APPROVED** [review]

File: packages/tooling/fast-tooling/src/web-components/html-render/html-render.ts [hunk 3]
```diff
@@ -105,52 +122,67 @@ export class HTMLRender extends FASTElement {
...
+        this.selectTimeout = null;
+        let el: HTMLElement = this.shadowRoot.querySelector(
             "[" + this.dataDictionaryAttr + "=" + this.activeDictionaryId + "]"
         );
+        while (!el && this.dataDictionary[0][this.activeDictionaryId].parent) {
+            this.activeDictionaryId = this.dataDictionary[0][
+                this.activeDictionaryId
+            ].parent.id;
+            el = this.shadowRoot.querySelector(
+                "[" + this.dataDictionaryAttr + "=" + this.activeDictionaryId + "]"
```
Comment (janechu) [inline packages/tooling/fast-tooling/src/web-components/html-render/html-render.ts:147]: nit: could use backticks here.

File: sites/fast-creator/app/preview.tsx [hunk 4]
```diff
@@ -168,14 +169,35 @@ class Preview extends Foundation<{}, {}, PreviewState> {
...
+                    options: {
+                        originatorId: "preview",
+                    },
+                    activeNavigationConfigId: "",
+                });
+            }
         }
     }
+
+    private attachComponentsAndData(): void {
```
Comment (janechu) [inline sites/fast-creator/app/preview.tsx:197]: Not sure I'm seeing where this is getting used.

File: packages/tooling/fast-tooling/src/web-components/html-render-layer-inline-edit/html-render-layer-inline-edit.ts (hunk 2 above)
Comment (williamw2) [reply packages/tooling/fast-tooling/src/web-components/html-render-layer-inline-edit/html-render-layer-inline-edit.ts]: Fixed

File: packages/tooling/fast-tooling/src/web-components/html-render-layer-inline-edit/html-render-layer-inline-edit.ts [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+            }
+            this.timeoutRef = window.setTimeout(() => {
+                this.applySizeAndPositionToTextbox();
+            }, 40);
+        }
+    };
+
+    public handleKeyDown(e: KeyboardEvent) {
+        if (e.key.length === 1) {
+            this.currentTextNode.textContent += e.key === " " ? " " : e.key;
```
[the remark this answers is not in the document]
Comment (williamw2) [reply packages/tooling/fast-tooling/src/web-components/html-render-layer-inline-edit/html-render-layer-inline-edit.ts]: This was leftover code from attempts to fix the whitespace issue that didn't get removed. It has been removed.

File: packages/tooling/fast-tooling/src/web-components/html-render-layer-inline-edit/html-render-layer-inline-edit.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (williamw2) [reply packages/tooling/fast-tooling/src/web-components/html-render-layer-inline-edit/html-render-layer-inline-edit.ts]: Fixed

File: packages/tooling/fast-tooling/src/web-components/html-render-layer-inline-edit/html-render-layer-inline-edit.ts [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+        const range = document.createRange();
+        range.selectNode(target);
+        const pos: DOMRect = range.getBoundingClientRect();
+        return new OverylayPosition(pos.top, pos.left, pos.width, pos.height);
+    }
+
+    private applySizeAndPositionToTextbox() {
+        this.textPosition = this.GetPositionFromElement(this.currentTextNode);
+        this.textAreaRef.style.top = this.textPosition.top + "px";
+        this.textAreaRef.style.left = this.textPosition.left + "px";
```
[the remark this answers is not in the document]
Comment (williamw2) [reply packages/tooling/fast-tooling/src/web-components/html-render-layer-inline-edit/html-render-layer-inline-edit.ts]: Fixed

File: packages/tooling/fast-tooling/src/web-components/html-render-layer-inline-edit/html-render-layer-inline-edit.ts [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+        });
+        this.currentDataId = null;
+        this.currentTextNode = null;
+        if (this.activityCallback) {
+            this.activityCallback(this.layerActivityId, ActivityType.releaseFocus);
+        }
+    }
+
+    private cancelEdit() {
+        // undo everything
```
[the remark this answers is not in the document]
Comment (williamw2) [reply packages/tooling/fast-tooling/src/web-components/html-render-layer-inline-edit/html-render-layer-inline-edit.ts]: Changed

File: packages/tooling/fast-tooling/src/web-components/html-render-layer-inline-edit/html-render-layer-inline-edit.ts [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+        this.currentTextNode = null;
+        this.textValue = "";
+        this.originalTextValue = "";
+        if (this.activityCallback) {
+            this.activityCallback(this.layerActivityId, ActivityType.update);
+            this.activityCallback(this.layerActivityId, ActivityType.releaseFocus);
+        }
+    }
+
+    public elementActivity(
```
[the remark this answers is not in the document]
Comment (williamw2) [reply packages/tooling/fast-tooling/src/web-components/html-render-layer-inline-edit/html-render-layer-inline-edit.ts:192]: Added

File: packages/tooling/fast-tooling/src/web-components/html-render/html-render.ts (hunk 3 above)
Comment (williamw2) [reply packages/tooling/fast-tooling/src/web-components/html-render/html-render.ts]: Fixed

File: sites/fast-creator/app/preview.tsx (hunk 4 above)
Comment (williamw2) [reply sites/fast-creator/app/preview.tsx]: It wasn't. Dead code that I failed to clean up. It has been removed.

File: packages/tooling/fast-tooling/src/web-components/html-render-layer-inline-edit/html-render-layer-inline-edit.ts [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+            i++;
+            this.currentStyleTarget = path[i] as HTMLElement;
+        }
+        // position, style and show the textarea
+        this.applyStylesToTextbox();
+        this.applySizeAndPositionToTextbox();
+        this.textAreaActive = true;
+        window.setTimeout(() => {
+            this.textAreaRef.focus();
+        }, 10);
```
[the remark this answers is not in the document]
Comment (williamw2) [reply packages/tooling/fast-tooling/src/web-components/html-render-layer-inline-edit/html-render-layer-inline-edit.ts:151]: Comment added. We need to give the textarea a chance to actually be shown before we can give it focus.

File: packages/tooling/fast-tooling/src/web-components/html-render-layer-navigation/html-render-layer-navigation.spec.ts [hunk 10]
```diff
@@ -84,7 +84,7 @@ xdescribe("HTMLRenderLayerNavgation", () => {
         const div = document.createElement("div");
         parent.appendChild(div);
 
-        element.elementActivity(ActivityType.click, "root", div);
+        element.elementActivity("test", ActivityType.click, "root", div);
```
[the remark this answers is not in the document]
Comment (williamw2) [reply packages/tooling/fast-tooling/src/web-components/html-render-layer-navigation/html-render-layer-navigation.spec.ts:87]: This case is simulating the element activity being sent to the navigation layer by the HTMLRender container. If anything it should be the HTMLRender's layerActivityId, however any string will do for this scenario.

File: packages/tooling/fast-tooling/src/web-components/html-render/html-render.template.ts [hunk 11]
```diff
@@ -9,6 +9,7 @@ export const HTMLRenderTemplate = html<HTMLRender>`
         <div
             class="html-render"
             @click="${(x, c) => x.clickHandler(c.event as MouseEvent)}"
+            @dblclick="${(x, c) => x.dblClickHandler(c.event as MouseEvent)}"
```
[the remark this answers is not in the document]
Comment (williamw2) [reply packages/tooling/fast-tooling/src/web-components/html-render/html-render.template.ts:12]: Yeah. It sends two click events followed by the double click. In our case we do want at least one of the clicks to happen in order to ensure that the correct element is selected before the double click attempts to locate a text node. The only downside is that we get a second unnecessary click event which doesn't cause any harm.

File: sites/fast-creator/app/web-components/index.tsx (hunk 1 above)
Comment (williamw2) [reply sites/fast-creator/app/web-components/index.tsx:323]: Yup. Adding new layers is easy.

**Review by janechu — APPROVED** [review]

**Pull request merged**